### PR TITLE
bpo-42129: Add support for resources in namespaces

### PR DESCRIPTION
**Pull request by jaraco**
And sync with [importlib_resources 5.0.3](https://importlib-resources.readthedocs.io/en/latest/history.html#v5-0-3).

https://bugs.python.org/issue42129

### CODE REVIEW

**Comment by jaraco** [comment]
Annoyingly, this change includes adoption of black in the backport, so includes all of the flux and losses in readability that brings.

**Comment by jaraco** [comment]
According to the test failures, something is modifying sys.path.

File: Lib/test/test_importlib/test_reader.py [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+        self.assertEqual(
+            repr(MultiplexedPath(self.folder)),
+            "MultiplexedPath('{}')".format(self.folder),
+        )
+
+
+class NamespaceReaderTest(unittest.TestCase):
+    @classmethod
+    def setUpClass(cls):
+        sys.path.append(os.path.abspath(os.path.join(__file__, '..')))
```
Comment (jaraco) [inline Lib/test/test_importlib/test_reader.py]: This test is likely the one that's violating the hermetic tests.

Comment (jaraco) [reply Lib/test/test_importlib/test_reader.py]: Likely addressed in 84008ab1ff.

**Comment by kousu** [comment]
Thank you so much for this!

I spent a long time this weekend trying to work around this for https://github.com/sct-data/spinalcordtoobox-data-template until I eventually just gave up.

**Pull request merged**